src/Microsoft.OData.Core/UriParser/Binders/InBinder.cs

```diff
@@ -97,21 +97,28 @@ private CollectionNode GetCollectionOperandFromToken(QueryToken queryToken, IEdm
             LiteralToken literalToken = queryToken as LiteralToken;
             if (literalToken != null)
             {
-                string originalLiteralText = literalToken.OriginalText;
-
                 // Parentheses-based collections are not standard JSON but bracket-based ones are.
                 // Temporarily switch our collection to bracket-based so that the JSON reader will
                 // correctly parse the collection. Then pass the original literal text to the token.
-                string bracketLiteralText = originalLiteralText;
-                if (bracketLiteralText[0] == '(')
+                string bracketLiteralText = literalToken.OriginalText;
+
+                if (bracketLiteralText[0] == '(' || bracketLiteralText[0] == '[')
                 {
-                    Debug.Assert(bracketLiteralText[bracketLiteralText.Length - 1] == ')',
-                        "Collection with opening '(' should have corresponding ')'");
+                    bool isParenthesesBased = bracketLiteralText[0] == '(' && bracketLiteralText[^1] == ')';
+                    bool isBracketBased = bracketLiteralText[0] == '[' && bracketLiteralText[^1] == ']';
+
+                    Debug.Assert(isParenthesesBased || isBracketBased,
+                        $"Collection with opening '{bracketLiteralText[0]}' should have corresponding '{bracketLiteralText[^1]}'");
 
-                    StringBuilder replacedText = new StringBuilder(bracketLiteralText);
-                    replacedText[0] = '[';
-                    replacedText[replacedText.Length - 1] = ']';
-                    bracketLiteralText = replacedText.ToString();
+                    if(isParenthesesBased)

```

Comment:
Can be simplified... No need for the variables, we only need to verify that the last character matches the first character:
```csharp
Debug.Assert(bracketLiteralText[0] == 'bracketLiteralText[^1],
    $"Collection with opening '{bracketLiteralText[0]}' should have corresponding '{(bracketLiteralText[0] == '(' ? ')' : ']')}'");

if (bracketLiteralText[0] == '(')
// ...
```

Note that with your current code, if `bracketLiteralText` had the value `(foobar]`, the logic would result into a debug assertion with the following message:
```
Collection with opening '(' should have corresponding ']'
```
Likewise, if `bracketLiteralText` had the value `(foobar!`, the logic would result into a debug assertion with the following message:
```
Collection with opening '(' should have corresponding '!'
```

Reply:
Thanks @gathogojr for pointing this out. 
Resolved.